File: libs/python/dsp_common/dsp_common/metrics.py
from __future__ import annotations

import http.server
import socketserver
import threading
from dataclasses import dataclass

from prometheus_client import CONTENT_TYPE_LATEST, CollectorRegistry, Counter, Histogram, generate_latest


@dataclass(frozen=True)
class Metrics:
    registry: CollectorRegistry
    http_requests_total: Counter
    http_request_duration_seconds: Histogram


def create_metrics(service_name: str) -> Metrics:
    """
    Metrics contract:
    - avoid unbounded label cardinality
    - label by route template, not raw path
    """
    reg = CollectorRegistry()
    http_requests_total = Counter(
        "http_requests_total",
        "Total HTTP requests",
        ["service", "method", "route", "status"],
        registry=reg,
    )
    http_request_duration_seconds = Histogram(
        "http_request_duration_seconds",
        "HTTP request duration in seconds",
        ["service", "method", "route", "status"],
        buckets=(0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10),
        registry=reg,
    )
    # The service label is a constant conceptually, but prometheus_client doesn't
    # support const labels natively on these primitives; enforce via wrapper usage.
    return Metrics(registry=reg, http_requests_total=http_requests_total, http_request_duration_seconds=http_request_duration_seconds)


def start_metrics_server(addr: str, metrics: Metrics) -> threading.Thread:
    """
    Start a minimal Prometheus scrape endpoint on /metrics.
    Uses stdlib HTTP server to avoid heavy dependencies.
    """

    host, port_str = addr.split(":") if ":" in addr else ("0.0.0.0", addr)
    port = int(port_str)

    class Handler(http.server.BaseHTTPRequestHandler):
        def do_GET(self) -> None:  # noqa: N802
            if self.path != "/metrics":
                self.send_response(404)
                self.end_headers()
                self.wfile.write(b"not found")
                return

            body = generate_latest(metrics.registry)
            self.send_response(200)
            self.send_header("Content-Type", CONTENT_TYPE_LATEST)
            self.end_headers()
            self.wfile.write(body)

        def log_message(self, format: str, *args) -> None:  # noqa: A002
            # Do not emit default unstructured access logs; services log explicitly.
            return

    httpd = socketserver.TCPServer((host, port), Handler)

    t = threading.Thread(target=httpd.serve_forever, name="metrics-server", daemon=True)
    t.start()
    return t

File: libs/typescript/dsp-common/src/metrics.ts
import http from "node:http";
import client from "prom-client";
import type { Logger } from "./logging.js";

/**
 * Metrics registry and helpers.
 *
 * Every service exposes:
 * - /metrics endpoint (Prometheus scrape)
 *
 * Services should:
 * - register default metrics
 * - define request counters/histograms
 * - avoid unbounded label cardinality
 */
export type Metrics = {
  registry: client.Registry;
  httpRequestsTotal: client.Counter<string>;
  httpRequestDurationSeconds: client.Histogram<string>;
};

export function createMetrics(serviceName: string): Metrics {
  const registry = new client.Registry();
  registry.setDefaultLabels({ service: serviceName });
  client.collectDefaultMetrics({ register: registry });

  const httpRequestsTotal = new client.Counter({
    name: "http_requests_total",
    help: "Total HTTP requests",
    labelNames: ["method", "route", "status"] as const,
    registers: [registry]
  });

  const httpRequestDurationSeconds = new client.Histogram({
    name: "http_request_duration_seconds",
    help: "HTTP request duration in seconds",
    labelNames: ["method", "route", "status"] as const,
    buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10],
    registers: [registry]
  });

  return { registry, httpRequestsTotal, httpRequestDurationSeconds };
}

export function startMetricsServer(opts: {
  addr: string;
  registry: client.Registry;
  logger: Logger;
}): http.Server {
  const [host, portStr] = opts.addr.includes(":") ? opts.addr.split(":") : ["0.0.0.0", opts.addr];
  const port = Number(portStr);
  if (!Number.isFinite(port)) {
    throw new Error(`Invalid metrics address: ${opts.addr}`);
  }

  const server = http.createServer(async (req, res) => {
    if (req.method === "GET" && req.url === "/metrics") {
      try {
        const body = await opts.registry.metrics();
        res.writeHead(200, { "Content-Type": opts.registry.contentType });
        res.end(body);
      } catch (err) {
        opts.logger.error({ err }, "metrics endpoint failed");
        res.writeHead(500, { "Content-Type": "text/plain" });
        res.end("metrics error");
      }
      return;
    }

    res.writeHead(404, { "Content-Type": "text/plain" });
    res.end("not found");
  });

  server.listen(port, host, () => {
    opts.logger.info({ addr: opts.addr }, "metrics server listening");
  });

  return server;
}

